tests/integration/test_optional_outputs.py

```diff
@@ -0,0 +1,476 @@
+
+        # the workflow should be stalled
+        assert schd.pool.is_stalled()
+
+        # this can get logged by two different interfaces
+        assert log_filter(
+            # automatically - when the task finishes
+            log,
+            contains="1/a did not complete required outputs: ['x', 'y']",
+        )
+        log.clear()
+        schd.pool.log_incomplete_tasks()
+        assert log_filter(
+            # via log_incomplete_tasks - called on shutdown
+            log,
+            contains="1/a did not complete required outputs: ['x', 'y']",
+        )
+
+        # set output x
+        schd.pool.spawn_on_output(task_a, 'x')
+
+        # the workflow should be unstalled
+        schd.is_stalled = None  # reset stall status
+        assert not schd.pool.is_stalled()
+
+
+async def test_workflow_stalls_if_no_opt_output_generated_2(
+    flow,
+    scheduler,
+    start,
+):
+    """It should stall if zero optional outputs are generated.
+
+    Example 2: succeeded  # but we actually want "succeeded or failed"
+
+    Because there is no failure pathway defined in this example, Cylc will
+    stall as it cannot determine the user's intention. To avoid this a failure
+    pathway must be provided in the graph, if there is no failure pathway
+    (i.e. `a:fail => null`) a completion expression can be written to tolerate
+    task failure.
+
+    Tests proposal point (2)
+    https://github.com/cylc/cylc-admin/blob/master/docs/proposal-optional-output-extension.md
+    """
+    raw = {
+        'scheduling': {
+            'graph': {
+                'R1': '''
+                    a? => x
+                '''
+            },
+        },
+        'runtime': {'a': {}, 'x': {}}
+    }
+
+    id_ = flow(raw)
+    schd = scheduler(id_)
+    async with start(schd):
+        # one or more optional outputs must be completed so, in the absense of
+        # an alternative pathway "succeeded" is required
+        task_a = schd.pool.get_task('1', 'a')
+        assert task_a.state.outputs._completion_expression == 'succeeded'
+
+        # if 1/a fails, the workflow should stall
+        task_a.state_reset(TASK_OUTPUT_FAILED)
+        assert schd.pool.is_stalled()
+
+    # to avoid stall in the event of failure the user must configure a failure
+    # pathway, either by `a:fail? => something` or by using a completion
+    # expression
+    raw['runtime']['a']['completion'] = 'succeeded or failed'
+
+    # the workflow should *not* stall if 1/a fails
+    id_ = flow(raw)
+    schd = scheduler(id_)
+    async with start(schd):
+        # Cylc now knows that failure is expected and ok
+        task_a = schd.pool.get_task('1', 'a')
+        assert task_a.state.outputs._completion_expression == (
+            'succeeded or failed'
+        )
+
+        # so the workflow should not stall if 1/a fails
+        task_a.state_reset(TASK_OUTPUT_FAILED)
+        assert not schd.pool.is_stalled()
+
+
+async def test_completion_expiression(
```

Comment:
```suggestion
async def test_completion_expression(
```

Reply:
I really wish "expiression" was a word though.

Reply:
Should have got this testing expiry!